envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoyListenersFactory.kt

```diff
@@ -0,0 +1,304 @@
+                .setHeader(header)
+                .setRemove(false)
+                .setOnHeaderPresent(
+                        HeaderToMetadataConfig.KeyValuePair.newBuilder()
+                                .setKey(tag)
+                                .setMetadataNamespace("envoy.lb")
+                                .setType(HeaderToMetadataConfig.ValueType.STRING)
+                                .build()
+                )
+                .build())
+    }
+
+    fun headerToMetadataConfig(
+        rules: List<HeaderToMetadataConfig.Rule>,
+        key: String = snapshotProperties.loadBalancing.canary.metadataKey
+    ): HeaderToMetadataConfig.Builder {
+        val headerToMetadataConfig = HeaderToMetadataConfig.newBuilder()
+                .addRequestRules(
+                        HeaderToMetadataConfig.Rule.newBuilder()
+                                .setHeader("x-canary")
+                                .setRemove(false)
+                                .setOnHeaderPresent(
+                                        HeaderToMetadataConfig.KeyValuePair.newBuilder()
+                                                .setKey(key)
+                                                .setMetadataNamespace("envoy.lb")
+                                                .setType(HeaderToMetadataConfig.ValueType.STRING)
+                                                .build()
+                                )
+                                .build()
+                )
+
+        rules.forEach {
+            headerToMetadataConfig.addRequestRules(it)
+        }
+
+        return headerToMetadataConfig
+    }
+
+    private fun envoyRouterHttpFilter(): HttpFilter = HttpFilter.newBuilder().setName("envoy.router").build()
+
+    fun headerToMetadataHttpFilter(headerToMetadataConfig: HeaderToMetadataConfig.Builder): HttpFilter {
+        return HttpFilter.newBuilder()
+                .setName("envoy.filters.http.header_to_metadata")
+                .setTypedConfig(ProtobufAny.pack(
+                        headerToMetadataConfig.build()
+                ))
+                .build()
+    }
+}
+
+@Suppress("MagicNumber")
+class EnvoyListenersFactory(
+    snapshotProperties: SnapshotProperties,
+    private val ingressFilters: List<HttpFilterFactory> = listOf(),
+    private val egressFilters: List<HttpFilterFactory> = listOf()
+) {
+    private val listenersFactoryProperties = snapshotProperties.dynamicListeners
+    private val accessLogTimeFormat = stringValue(listenersFactoryProperties.httpFilters.accessLog.timeFormat)
+    private val accessLogMessageFormat = stringValue(listenersFactoryProperties.httpFilters.accessLog.messageFormat)
+    private val accessLogLevel = stringValue(listenersFactoryProperties.httpFilters.accessLog.level)
+    private val accessLogLogger = stringValue(listenersFactoryProperties.httpFilters.accessLog.logger)
+    private val egressRdsInitialFetchTimeout: Long = 20
```

Comment:
It wasn't obvious to me that these timeouts are in seconds. I'm not aware of any convention in EC but from what I see most timeouts use `Duration`. Actually the only place these are used they are immediately wrapped.
```suggestion
    private val egressRdsInitialFetchTimeout: Duration = durationInSeconds(20)
```